packages/utils/node-resolver-rs/src/lib.rs

```diff
@@ -0,0 +1,2392 @@
+      module_dir_resolver: None,
+    }
+  }
+
+  pub fn resolve<'s>(
+    &self,
+    specifier: &'s str,
+    from: &Path,
+    specifier_type: SpecifierType,
+  ) -> ResolveResult<'s> {
+    let invalidations = Invalidations::default();
+    let (specifier, query) = match Specifier::parse(specifier, specifier_type, self.flags) {
+      Ok(s) => s,
+      Err(e) => {
+        return ResolveResult {
+          result: Err(e.into()),
+          invalidations,
+        }
+      }
+    };
+    let request = ResolveRequest::new(self, &specifier, specifier_type, from, &invalidations);
+    let result = match request.resolve() {
+      Ok(r) => Ok((r, query)),
+      Err(r) => Err(r),
+    };
+
+    ResolveResult {
+      result,
+      invalidations,
+    }
+  }
+
+  pub fn resolve_side_effects(&self, path: &Path) -> Result<bool, ResolverError> {
+    if let Some(package) = self.find_package(path.parent().unwrap())? {
+      Ok(package.has_side_effects(path))
+    } else {
+      Ok(true)
+    }
+  }
+
+  fn find_package(&self, from: &Path) -> Result<Option<&PackageJson>, ResolverError> {
+    if let Some(path) = self.find_ancestor_file(from, "package.json") {
+      let package = self.cache.read_package(Cow::Owned(path))?;
+      return Ok(Some(package));
+    }
+
+    Ok(None)
+  }
+
+  fn find_ancestor_file(&self, from: &Path, filename: &str) -> Option<PathBuf> {
+    for dir in from.ancestors() {
+      if let Some(filename) = dir.file_name() {
+        if filename == "node_modules" {
+          break;
+        }
+      }
+
+      let file = dir.join(filename);
+      if self.cache.fs.is_file(&file) {
+        return Some(file);
+      }
+
+      if dir == self.project_root {
+        break;
+      }
+    }
+
+    None
+  }
+}
+
+struct ResolveRequest<'a, Fs> {
+  resolver: &'a Resolver<'a, Fs>,
+  specifier: &'a Specifier<'a>,
+  specifier_type: SpecifierType,
+  from: &'a Path,
+  flags: RequestFlags,
+  tsconfig: OnceCell<Option<&'a TsConfig<'a>>>,
+  invalidations: &'a Invalidations,
+  conditions: ExportsCondition,
+  priority_extension: Option<&'a str>,
+}
+
+bitflags! {
+  struct RequestFlags: u8 {
+    const IN_TS_FILE = 1 << 0;
+    const IN_JS_FILE = 1 << 1;
+    const IN_NODE_MODULES = 1 << 2;
+  }
+}
+
+impl<'a, Fs: FileSystem> ResolveRequest<'a, Fs> {
+  fn new(
+    resolver: &'a Resolver<'a, Fs>,
+    specifier: &'a Specifier<'a>,
+    mut specifier_type: SpecifierType,
+    from: &'a Path,
+    invalidations: &'a Invalidations,
+  ) -> Self {
+    let mut flags = RequestFlags::empty();
+    if let Some(ext) = from.extension() {
+      if ext == "ts" || ext == "tsx" || ext == "mts" || ext == "cts" {
+        flags |= RequestFlags::IN_TS_FILE;
+      } else if ext == "js" || ext == "jsx" || ext == "mjs" || ext == "cjs" {
+        flags |= RequestFlags::IN_JS_FILE;
+      }
+    }
+
+    if from.components().any(|c| c.as_os_str() == "node_modules") {
+      flags |= RequestFlags::IN_NODE_MODULES;
```

Comment:
This might not apply with yarn pnp?

Reply:
A package named `sub` might have this filepath with PnP: `/Users/niklas/.yarn/berry/cache/sub-file-8e8024d312-8.zip/node_modules/sub/index.js`. So there will be a node_modules in there (if that's what you were concerned about)